File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/calendar/HomeCalendarDerivedItem.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.ui.screens.homes.calendar

import androidx.compose.ui.graphics.Color
import app.pantopus.android.data.api.models.homes.BillDto
import app.pantopus.android.data.api.models.homes.HomeTaskDto
import app.pantopus.android.data.api.models.homes.PackageDto
import app.pantopus.android.ui.screens.homes.HomeDashboardProjection
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon

/**
 * Which feed a derived calendar row came from. The tone matches the Home
 * dashboard's own per-feature accents (tasks = warning, bills = error,
 * deliveries = business) so the colour-coding reads the same across the
 * pillar.
 *
 * Parity contract — mirrored in iOS `HomeCalendarDerivedKind`.
 */
enum class HomeCalendarDerivedKind(
    val label: String,
    val icon: PantopusIcon,
    val background: Color,
    val foreground: Color,
) {
    Task("Task", PantopusIcon.ListChecks, PantopusColors.warningBg, PantopusColors.warning),
    Bill("Bill", PantopusIcon.Receipt, PantopusColors.errorBg, PantopusColors.error),
    Package("Delivery", PantopusIcon.Package, PantopusColors.businessBg, PantopusColors.business),
}

/**
 * One derived, dated row for the Home calendar — a task due-date, a bill
 * due-date, or a package's expected delivery.
 *
 * RN's month grid plots exactly these alongside the home's own events
 * (`src/app/homes/[id]/calendar.tsx:48-74`): it fans out to
 * `getHomeTasks` / `getHomeBills` / `getHomePackages`, keeps only rows
 * that carry a date, and colour-codes each by type. Without it a
 * household sees an empty-looking month even when three bills are due
 * that week.
 */
data class HomeCalendarDerivedItem(
    val id: String,
    val kind: HomeCalendarDerivedKind,
    val title: String,
    /** Short status / amount line rendered under the title. */
    val detail: String?,
    /**
     * The date this item lands on — `due_at` (task), `due_date` (bill)
     * or `expected_at` / `delivered_at` (package).
     */
    val dateIso: String,
) {
    companion object {
        /**
         * Fan-in from the three feeds. Rows without a date are dropped —
         * they have nothing to plot. Mirrors RN's `if (t.due_date)` /
         * `if (b.due_date)` / `p.expected_date || p.delivered_at` guards.
         */
        fun build(
            tasks: List<HomeTaskDto>,
            bills: List<BillDto>,
            packages: List<PackageDto>,
        ): List<HomeCalendarDerivedItem> =
            buildList {
                tasks.forEach { task ->
                    val due = task.dueAt
                    if (!due.isNullOrEmpty()) {
                        add(
                            HomeCalendarDerivedItem(
                                id = "task-${task.id}",
                                kind = HomeCalendarDerivedKind.Task,
                                title = task.title,
                                detail = statusDetail(task.status),
                                dateIso = due,
                            ),
                        )
                    }
                }
                bills.forEach { bill ->
                    val due = bill.dueDate
                    if (!due.isNullOrEmpty()) {
                        add(
                            HomeCalendarDerivedItem(
                                id = "bill-${bill.id}",
                                kind = HomeCalendarDerivedKind.Bill,
                                title = "${HomeDashboardProjection.billLabel(bill)} bill due",
                                detail = HomeDashboardProjection.currency(bill.displayAmount, bill.currency),
                                dateIso = due,
                            ),
                        )
                    }
                }
                packages.forEach { parcel ->
                    val date =
                        listOfNotNull(parcel.expectedAt, parcel.deliveredAt)
                            .firstOrNull { it.isNotEmpty() }
                    if (date != null) {
                        val name =
                            listOfNotNull(parcel.description, parcel.vendorName, parcel.carrier)
                                .firstOrNull { it.isNotEmpty() } ?: "Package"
                        add(
                            HomeCalendarDerivedItem(
                                id = "package-${parcel.id}",
                                kind = HomeCalendarDerivedKind.Package,
                                title = name,
                                detail = statusDetail(parcel.status),
                                dateIso = date,
                            ),
                        )
                    }
                }
            }

        /**
         * `in_progress` → "In progress". Null for empty statuses so the
         * row falls back to its type label.
         */
        fun statusDetail(raw: String?): String? {
            if (raw.isNullOrEmpty()) return null
            return raw.replace('_', ' ').replaceFirstChar { it.uppercase() }
        }
    }
}

File: frontend/apps/ios/Pantopus/Features/Homes/Calendar/HomeCalendarDerivedItem.swift
//
//  HomeCalendarDerivedItem.swift
//  Pantopus
//
//  Task due-dates, bill due-dates and package expected-delivery dates
//  plotted on the Home calendar alongside the home's own events.
//
//  RN's month grid does exactly this
//  (`src/app/homes/[id]/calendar.tsx:48-74`): it fans out to
//  `getHomeTasks` / `getHomeBills` / `getHomePackages`, keeps only the
//  rows that carry a date, and colour-codes each by type. Without it a
//  household sees an empty-looking month even when three bills are due
//  that week.
//
//  Parity contract — mirrored in Android
//  `ui/screens/homes/calendar/HomeCalendarDerivedItem.kt`.
//

import Foundation
import SwiftUI

/// Which feed a derived calendar row came from. The tone matches the
/// Home dashboard's own per-feature accents (tasks = warning,
/// bills = error, deliveries = business) so the colour-coding reads the
/// same across the pillar.
public enum HomeCalendarDerivedKind: String, Sendable, Hashable, CaseIterable {
    case task
    case bill
    case package

    /// Chip / section label.
    public var label: String {
        switch self {
        case .task: "Task"
        case .bill: "Bill"
        case .package: "Delivery"
        }
    }

    public var icon: PantopusIcon {
        switch self {
        case .task: .listChecks
        case .bill: .receipt
        case .package: .package
        }
    }

    public var background: Color {
        switch self {
        case .task: Theme.Color.warningBg
        case .bill: Theme.Color.errorBg
        case .package: Theme.Color.businessBg
        }
    }

    public var foreground: Color {
        switch self {
        case .task: Theme.Color.warning
        case .bill: Theme.Color.error
        case .package: Theme.Color.business
        }
    }
}

/// One derived, dated row for the Home calendar.
public struct HomeCalendarDerivedItem: Sendable, Hashable, Identifiable {
    public let id: String
    public let kind: HomeCalendarDerivedKind
    public let title: String
    /// Short status / amount line rendered under the title.
    public let detail: String?
    /// The date this item lands on — `due_at` (task), `due_date` (bill)
    /// or `expected_at` / `delivered_at` (package).
    public let dateISO: String

    public init(
        id: String,
        kind: HomeCalendarDerivedKind,
        title: String,
        detail: String?,
        dateISO: String
    ) {
        self.id = id
        self.kind = kind
        self.title = title
        self.detail = detail
        self.dateISO = dateISO
    }

    /// Fan-in from the three feeds. Rows without a date are dropped —
    /// they have nothing to plot. Mirrors RN's `if (t.due_date)` /
    /// `if (b.due_date)` / `const d = p.expected_date || p.delivered_at`
    /// guards.
    public static func build(
        tasks: [HomeTaskDTO],
        bills: [BillDTO],
        packages: [PackageDTO]
    ) -> [HomeCalendarDerivedItem] {
        var out: [HomeCalendarDerivedItem] = []

        for task in tasks {
            guard let due = task.dueAt, !due.isEmpty else { continue }
            out.append(
                HomeCalendarDerivedItem(
                    id: "task-\(task.id)",
                    kind: .task,
                    title: task.title,
                    detail: statusDetail(task.status),
                    dateISO: due
                )
            )
        }

        for bill in bills {
            guard let due = bill.dueDate, !due.isEmpty else { continue }
            let name = HomeDashboardProjection.billLabel(bill)
            out.append(
                HomeCalendarDerivedItem(
                    id: "bill-\(bill.id)",
                    kind: .bill,
                    title: "\(name) bill due",
                    detail: HomeDashboardProjection.currency(
                        bill.displayAmount,
                        code: bill.currency
                    ),
                    dateISO: due
                )
            )
        }

        for parcel in packages {
            let date = [parcel.expectedAt, parcel.deliveredAt]
                .compactMap { $0 }
                .first { !$0.isEmpty }
            guard let date else { continue }
            let name = [parcel.description, parcel.vendorName, parcel.carrier]
                .compactMap { $0 }
                .first { !$0.isEmpty } ?? "Package"
            out.append(
                HomeCalendarDerivedItem(
                    id: "package-\(parcel.id)",
                    kind: .package,
                    title: name,
                    detail: statusDetail(parcel.status),
                    dateISO: date
                )
            )
        }

        return out
    }

    /// `in_progress` → "In progress". Nil for empty statuses so the row
    /// falls back to its type label.
    static func statusDetail(_ raw: String?) -> String? {
        guard let raw, !raw.isEmpty else { return nil }
        let spaced = raw.replacingOccurrences(of: "_", with: " ")
        guard let first = spaced.first else { return nil }
        return first.uppercased() + spaced.dropFirst()
    }

    /// Row projection for the agenda list. Read-only — these rows are a
    /// mirror of surfaces that own their own screens, so tapping is a
    /// no-op exactly as in RN's calendar.
    /// `@MainActor` because the date helpers it borrows from
    /// `HomeCalendarViewModel` are main-actor isolated; the only caller is the
    /// view model itself, which is already on the main actor.
    @MainActor
    public func row(calendar: Calendar) -> RowModel {
        let timeLabel = HomeCalendarViewModel.parseIsoInstant(dateISO).map {
            HomeCalendarViewModel.formatTime(start: $0, endIso: nil, calendar: calendar)
        }
        return RowModel(
            id: id,
            title: title,
            subtitle: detail ?? kind.label,
            template: .statusChip,
            leading: .typeIcon(
                kind.icon,
                background: kind.background,
                foreground: kind.foreground
            ),
            trailing: .none,
            chips: [
                RowChip(
                    text: kind.label,
                    icon: kind.icon,
                    tint: .custom(
                        background: kind.background,
                        foreground: kind.foreground
                    )
                )
            ],
            timeMeta: timeLabel
        )
    }
}
